### PR TITLE
Yairp/feat/call webhooks

### PR DESCRIPTION
**Pull request by yairper**
### Description
Elaborate on the subject, motivation, and context.

### Related issues
 * Provide a link to each related issue.

### Breaking changes
 * Describe the breaking changes that this pull request introduces.

### How these changes were tested
 * Describe the tests that you ran to verify your changes, including devices, operating systems, browsers and versions.

### Examples and references
 * Links, screenshots, and other resources related to this change.

### Checklist
- [] I have read the [contribution guidelines](CONTRIBUTING.md) of this project
- [] I have read the [style guidelines](STYLE_GUIDE.md) of this project
- [] I have performed a self-review of my own code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings and errors
- [] New and existing tests pass locally with my changes

### CODE REVIEW

File: services/workflows-service/src/workflow/context-changed-webhook-caller.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    const oldDocuments = runtimeData.context['documents'] || [];
+    const newDocuments = newContext['documents'] || [];
+
+    const anyDocumentStatusChanged = (oldDocuments as Array<any>).some(
+      (oldDocument, i) => oldDocument.decision.status !== newDocuments[i].decision.status,
+    );
+
+    const id = randomUUID();
+
+    setTimeout(() => {
```
Comment (alonp99) [inline services/workflows-service/src/workflow/context-changed-webhook-caller.ts:29]: Why setTimeout?

File: services/workflows-service/src/workflow/context-changed-webhook-caller.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      this.#__httpService.post(this.#__webhookConfig.url, {
+        id,
```
Comment (alonp99) [inline services/workflows-service/src/workflow/context-changed-webhook-caller.ts:31]: eventually, i think this should be extracted to a webhook service, he will hold the knowledge webhook data wrapper and other logic related to sending the webhook etc... - so we can handle additional events quickly (if a user asks to get also XYZ event) - and also can quickly extract it to another process (to improve reliability)

It's not an immediate need but I think its smart to have it so we can react quickly

File: services/workflows-service/src/workflow/context-changed-webhook-caller.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable @typescript-eslint/no-unsafe-return */
+
+import { randomUUID } from 'crypto';
+
+export class ContextChangedWebookCaller {
+  #__webhookConfig: ProcessEnvWebhookConfig;
+  #__httpService;
+
+  constructor(eventEmitter: any, webhookConfig: any, httpService: any) {
```
Comment (alonp99) [inline services/workflows-service/src/workflow/context-changed-webhook-caller.ts:10]: I still dont have a cleat picture in mind about what we win/lose with this approach,

I do know that we will lose nest eventEmitter features like events wildcards (`workflow.*`)

File: services/workflows-service/src/workflow/context-changed-webhook-caller.ts (hunk 2 above)
Comment (alonp99) [reply services/workflows-service/src/workflow/context-changed-webhook-caller.ts]: move to webhook service

File: services/workflows-service/src/workflow/context-changed-webhook-caller.ts (hunk 3 above)
Comment (yairper) [reply services/workflows-service/src/workflow/context-changed-webhook-caller.ts]: fixed

File: services/workflows-service/src/workflow/context-changed-webhook-caller.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      return (
+        id in newDocumentsByIdentifier &&
+        oldDocument.decision.status !== newDocumentsByIdentifier[id].decision.status
+      );
+    });
+
+    const id = randomUUID();
+
+    if (anyDocumentStatusChanged) {
+      setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (yairper) [reply services/workflows-service/src/workflow/context-changed-webhook-caller.ts:43]: fixed

File: services/workflows-service/src/events/event-consumer.ts [hunk 5]
```diff
@@ -1,21 +1,71 @@
...
-  subscribe(event: string, callback: ListenerFn) {
-    this.eventEmitter.addListener(event, eventData => {
-      callback(undefined, eventData);
+  @OnEvent('workflow.context.changed')
+  handleWorkflowEvent(data: WorkflowEventRawData) {
+    const oldDocuments = (data.runtimeData.context as any)['documents'] || [];
+    const newDocuments = data.context?.['documents'] || [];
+
+    const documentIdentifier = (doc: any) => {
+      return `${doc.category as string}$${doc.type as string}$${doc.issuer?.country as string}`;
```
Comment (alonp99) [inline services/workflows-service/src/events/event-consumer.ts:22]: change to `${doc.category as string}-${doc.type as string}-${doc.issuer?.country as string}`.toLowerCAse()

File: services/workflows-service/src/events/event-consumer.ts [hunk 6]
```diff
@@ -1,21 +1,71 @@
...
+            eventName: 'workflow.context.document.changed',
+            apiVersion: 1,
+            timestamp: new Date().toISOString(),
+            workflowRuntimeId: data.runtimeData.id,
+            environment,
+            data: data.context,
+          },
+          {
+            auth: {
+              username: 'TBD',
```
Comment (alonp99) [inline services/workflows-service/src/events/event-consumer.ts:61]: change to authorization header, with value `Secret XXX`

File: services/workflows-service/src/events/event-consumer.ts [hunk 7]
```diff
@@ -1,21 +1,71 @@
...
     });
-  }
 
-  // OnEvent is not working, need to check why, for now we subscribe explicitly
-  // @OnEvent('workflow.a')
-  // handleWorkflowEvent(eventName: string, data: WorkflowEventData) {
-  //   console.log('eeeee', eventName, data);
-  // }
+    const id = randomUUID();
+    const url = process.env.WEBHOOK_URL as string;
```
Comment (alonp99) [inline services/workflows-service/src/events/event-consumer.ts:40]: use `env`

**Pull request merged**